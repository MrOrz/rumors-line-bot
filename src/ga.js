import ua from 'universal-analytics';

// When document title is too long, event will be dropped.
// See: https://github.com/cofacts/rumors-line-bot/issues/97
//
const DOCUMENT_TITLE_LENGTH = 800;

/**
 * Sends a screen view and returns the visitor
 *
 * @param {string} userId
 * @param {string} state - The current state the LINE bot is in
 * @param {string} documentTitle - The article content (user input) currently in search
 * @returns {ua.Visitor} universal analytics visitor
 */
export default function ga(userId, state = 'N/A', documentTitle = '') {
  const visitor = ua(process.env.GA_ID, userId, { strictCidFormat: false });

  // Ref: https://developers.google.com/analytics/devguides/collection/analyticsjs/screens
  // screenview needs both screen name and application name, or it will print error
  //
  visitor.screenview(state, 'rumors-line-bot');

  if (documentTitle) {
    visitor.set('dt', documentTitle.slice(0, DOCUMENT_TITLE_LENGTH));
  }

  return visitor;
}
