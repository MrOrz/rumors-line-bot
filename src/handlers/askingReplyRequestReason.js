import gql from '../gql';
import {
  createPostbackAction,
  getArticleURL,
  REASON_PLACEHOLDER,
} from './utils';

export default async function askingArticleSubmission(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;
  const { selectedArticleId } = data;

  if (event.input !== 'n' && event.input !== REASON_PLACEHOLDER) {
    const reason = event.input;

    const text =
      `以下是您所填寫的理由：\n「\n${reason}\n」\n` +
      '我們即將把您填寫的理由送至資料庫。' +
      '若您送出的訊息或理由意味不明、造成闢謠編輯的困擾，可能會影響到您未來送出文章的權利。' +
      '若要確認請輸入「y」、若不想填寫請輸入「n」、若要重新填寫理由請輸入「r」';

    replies = [
      {
        type: 'flex',
        altText: text,
        contents: {
          type: 'bubble',
          styles: {
            footer: {
              separator: true,
            },
          },
          body: {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'text',
                text: '把您的理由送進資料庫',
                weight: 'bold',
                color: '#1DB446',
                size: 'sm',
              },
              {
                type: 'separator',
                margin: 'xxl',
              },
              {
                type: 'text',
                text: '您之前傳送的訊息',
                weight: 'bold',
                size: 'xl',
                margin: 'xl',
              },
              {
                type: 'text',
                text: data.selectedArticleText,
                size: 'xs',
              },
              {
                type: 'separator',
                margin: 'xxl',
              },
              {
                type: 'text',
                text: '覺得是謠言的理由',
                weight: 'bold',
                size: 'xl',
                margin: 'xl',
              },
              {
                type: 'text',
                text: reason,
                size: 'xxs',
              },
              {
                type: 'separator',
                margin: 'xxl',
              },
              {
                type: 'text',
                text:
                  '若您送出的理由意味不明、造成闢謠編輯的困擾，可能會影響到您未來送出文章的權利。',
                color: '#ff0000',
                wrap: true,
                margin: 'xxl',
              },
              {
                type: 'button',
                margin: 'xl',
                style: 'primary',
                action: createPostbackAction('我不想填了', 'n', issuedAt),
              },
              {
                type: 'button',
                action: createPostbackAction('重寫送出的理由', 'r', issuedAt),
              },
              {
                type: 'button',
                action: createPostbackAction('明白我要送出', 'y', issuedAt),
              },
            ],
          },
        },
      },
    ];
    data.reasonText = reason;
    state = 'ASKING_REPLY_REQUEST_SUBMISSION';
  } else {
    const {
      data: { CreateReplyRequest },
    } = await gql`
      mutation($id: String!) {
        CreateReplyRequest(articleId: $id) {
          replyRequestCount
        }
      }
    `({ id: selectedArticleId }, { userId });

    replies = [
      {
        type: 'text',
        text: `已經將您的需求記錄下來了，共有 ${
          CreateReplyRequest.replyRequestCount
        } 人跟您一樣渴望看到針對這篇訊息的回應。若有最新回應，會寫在這個地方：${getArticleURL(
          selectedArticleId
        )}`,
      },
    ];
    state = '__INIT__';
  }

  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
