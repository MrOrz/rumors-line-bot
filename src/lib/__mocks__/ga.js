export const eventMock = jest.fn();

function ga() {
  return {
    event: eventMock,
    send: () => {},
    screenview: () => {},
  };
}

export default ga;
