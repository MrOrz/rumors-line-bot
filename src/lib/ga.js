import ua from 'universal-analytics';
import GraphemeSplitter from 'grapheme-splitter';
const splitter = new GraphemeSplitter();

// When document title is too long, event will be dropped.
// See: https://github.com/cofacts/rumors-line-bot/issues/97
//
const DOCUMENT_TITLE_LENGTH = 800;

/**
 * Sends a screen view and returns the visitor
 *
 * @param {string} uuid - Visitor uuid, corresponding to `messageSource`, it will be userId, groupId, roomId
 * @param {string} state - The current state the LINE bot is in
 * @param {string} documentTitle - The article content (user input) currently in search
 * @param {string} messageSource - The event source { 'user' | 'group' | 'room' }
 * @returns {ua.Visitor} universal analytics visitor
 */
export default function ga(
  uuid,
  state = 'N/A',
  documentTitle = '',
  messageSource = 'user'
) {
  const visitor = ua(process.env.GA_ID, uuid, { strictCidFormat: false });

  // Ref: https://developers.google.com/analytics/devguides/collection/analyticsjs/screens
  // screenview needs both screen name and application name, or it will print error
  //
  visitor.screenview(state, 'rumors-line-bot');

  if (documentTitle) {
    visitor.set(
      'dt',
      splitter
        .splitGraphemes(documentTitle)
        .slice(0, DOCUMENT_TITLE_LENGTH)
        .join('')
    );
  }

  visitor.set('cd1', messageSource);

  return visitor;
}
