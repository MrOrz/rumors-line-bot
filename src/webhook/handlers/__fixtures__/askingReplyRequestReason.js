export const createReplyRequestSuccess = {
  data: {
    CreateOrUpdateReplyRequest: {
      replyRequestCount: 1,
    },
  },
};
