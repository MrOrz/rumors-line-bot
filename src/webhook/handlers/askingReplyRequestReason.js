import gql from '../../lib/gql';
import { getArticleURL, REASON_PREFIX, createArticleShareReply } from './utils';

export default async function askingArticleSubmission(params) {
  let { data, state, event, issuedAt, userId, replies, isSkipUser } = params;
  const { selectedArticleId } = data;

  if (!event.input.startsWith(REASON_PREFIX)) {
    replies = [
      {
        type: 'text',
        text: '請點擊上面的「我也想知道」，或放棄這則，改轉傳其他訊息。',
      },
    ];
  } else {
    const reason = event.input.slice(REASON_PREFIX.length);

    const {
      data: { CreateOrUpdateReplyRequest },
    } = await gql`
      mutation($id: String!, $reason: String) {
        CreateOrUpdateReplyRequest(articleId: $id, reason: $reason) {
          replyRequestCount
        }
      }
    `({ id: selectedArticleId, reason }, { userId });

    const articleUrl = getArticleURL(selectedArticleId);

    replies = [
      {
        type: 'text',
        text: `已經將您的需求記錄下來了，共有 ${
          CreateOrUpdateReplyRequest.replyRequestCount
        } 人跟您一樣渴望看到針對這篇訊息的回應。若有最新回應，會寫在這個地方：${articleUrl}`,
      },
      createArticleShareReply(articleUrl, reason),
    ];
    state = '__INIT__';
  }

  return { data, state, event, issuedAt, userId, replies, isSkipUser };
}
