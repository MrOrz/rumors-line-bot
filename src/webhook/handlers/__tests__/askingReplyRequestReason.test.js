jest.mock('src/lib/ga');

import askingReplyRequestReason from '../askingReplyRequestReason';
import {
  REASON_PREFIX,
  SOURCE_PREFIX,
  ARTICLE_SOURCE_OPTIONS,
} from 'src/lib/sharedUtils';
import ga from 'src/lib/ga';

beforeEach(() => {
  ga.clearAllMocks();
});

it('should block incorrect prefix', async () => {
  const params = {
    data: {
      selectedArticleId: 'selected-article-id',
    },
    state: 'ASKING_REPLY_REQUEST_REASON',
    event: {
      type: 'message',
      input: REASON_PREFIX + 'foo', // Wrong prefix
    },
  };
  await expect(askingReplyRequestReason(params)).rejects.toMatchInlineSnapshot(
    `[Error: Please press the latest button to submit message to database.]`
  );
});

it('records article source', async () => {
  const params = {
    data: {
      selectedArticleId: 'selected-article-id',
    },
    state: 'ASKING_REPLY_REQUEST_REASON',
    event: {
      input:
        SOURCE_PREFIX + ARTICLE_SOURCE_OPTIONS.find(({ valid }) => valid).label, // From LIFF
    },
  };
  expect(await askingReplyRequestReason(params)).toMatchSnapshot();
  expect(ga.eventMock.mock.calls).toMatchInlineSnapshot(`
    Array [
      Array [
        Object {
          "ea": "ProvidingSource",
          "ec": "UserInput",
          "el": "group message",
        },
      ],
      Array [
        Object {
          "ea": "ProvidingSource",
          "ec": "Article",
          "el": "selected-article-id/group message",
        },
      ],
    ]
  `);
  expect(ga.sendMock).toHaveBeenCalledTimes(1);
});
