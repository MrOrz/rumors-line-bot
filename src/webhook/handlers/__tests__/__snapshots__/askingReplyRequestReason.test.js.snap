// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`records article source 1`] = `
Object {
  "data": Object {
    "selectedArticleId": "selected-article-id",
  },
  "event": Object {
    "input": "‚ÑπÔ∏è I got the message from:
A LINE group",
  },
  "isSkipUser": undefined,
  "issuedAt": undefined,
  "replies": Array [
    Object {
      "altText": "Thanks for the info.",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "Thanks for the info.",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "footer": Object {
          "contents": Array [
            Object {
              "action": Object {
                "label": "See reported message",
                "type": "uri",
                "uri": "https://cofacts.hacktabl.org/article/selected-article-id",
              },
              "color": "#333333",
              "style": "primary",
              "type": "button",
            },
            Object {
              "action": Object {
                "label": "Provide more info",
                "type": "uri",
                "uri": "line://app/1631030389-xb5mnDdN?p=reason&token=eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE1ODc3NTI0MjguODg5fQ.jO37_Cd6Y4EZoTa6j4VaS8d04P2yHP2NEP8p_pa72PM",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "spacing": "sm",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
    Object {
      "altText": "Your friends may know the answer üåü Share your question to friends, maybe someone can help!",
      "contents": Object {
        "body": Object {
          "contents": Array [
            Object {
              "text": "Your friends may know the answer üåü Share your question to friends, maybe someone can help!",
              "type": "text",
              "wrap": true,
            },
          ],
          "layout": "vertical",
          "type": "box",
        },
        "footer": Object {
          "contents": Array [
            Object {
              "action": Object {
                "label": "Share on LINE",
                "type": "uri",
                "uri": "line://msg/text/?Please%20help%20me%20verify%20if%20this%20is%20true%3A%20https%3A%2F%2Fcofacts.hacktabl.org%2Farticle%2Fselected-article-id",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
            Object {
              "action": Object {
                "label": "Share on Facebook",
                "type": "uri",
                "uri": "https://www.facebook.com/dialog/share?openExternalBrowser=1&app_id=719656818195367&display=popup&hashtag=%23ReportedToCofacts&href=https%3A%2F%2Fcofacts.hacktabl.org%2Farticle%2Fselected-article-id",
              },
              "color": "#ffb600",
              "style": "primary",
              "type": "button",
            },
          ],
          "layout": "vertical",
          "spacing": "sm",
          "type": "box",
        },
        "type": "bubble",
      },
      "type": "flex",
    },
  ],
  "state": "__INIT__",
  "userId": undefined,
}
`;
